src/test/java/org/jenkinsci/plugins/workflow/job/CpsPersistenceTest.java

```diff
@@ -0,0 +1,432 @@
+            assertCompletedCleanly(run);
+        });
+    }
+
+    @Test
+    /** Build okay but program fails to load */
+    public void inProgressButProgramLoadFailure() throws Exception {
+        final int[] build = new int[1];
+        story.thenWithHardShutdown( j -> {
+            WorkflowRun run = runBasicPauseOnInput(j, DEFAULT_JOBNAME, build);
+            CpsFlowExecution cpsExec = (CpsFlowExecution)(run.getExecution());
+            Method m = cpsExec.getClass().getDeclaredMethod("getProgramDataFile", null);
+            m.setAccessible(true);
+            File f = (File)(m.invoke(cpsExec, null));
+            f.delete();
+        });
+        story.then( j->{
+            WorkflowJob r = j.jenkins.getItemByFullName(DEFAULT_JOBNAME, WorkflowJob.class);
+            WorkflowRun run = r.getBuildByNumber(build[0]);
+            assertCompletedCleanly(run);
+        });
+    }
+
+    @Test
+    /** Build okay but then the start nodes get screwed up */
+    public void inProgressButStartBlocksLost() throws Exception {
+        final int[] build = new int[1];
+        story.thenWithHardShutdown( j -> {
+            WorkflowRun run = runBasicPauseOnInput(j, DEFAULT_JOBNAME, build);
+            CpsFlowExecution cpsExec = (CpsFlowExecution)(run.getExecution());
+            getCpsBlockStartNodes(cpsExec).push(new FlowStartNode(cpsExec, cpsExec.iotaStr()));
+            run.save();
+        });
+        story.then( j->{
+            WorkflowJob r = j.jenkins.getItemByFullName(DEFAULT_JOBNAME, WorkflowJob.class);
+            WorkflowRun run = r.getBuildByNumber(build[0]);
+            assertCompletedCleanly(run);
+        });
+    }
+
+    @Test
+    @Issue("JENKINS-50199")
+    /** Replicates case where builds resume when the should not due to build's completion not being saved. */
+    public void completedExecutionButRunIncomplete() throws Exception {
+        final int[] build = new int[1];
+        story.thenWithHardShutdown( j -> {
+            WorkflowRun run = runBasicPauseOnInput(j, DEFAULT_JOBNAME, build);
+            CpsFlowExecution cpsExec = (CpsFlowExecution)(run.getExecution());
+            InputStepExecution exec = getInputStepExecution(run, "pause");
+            exec.doProceedEmpty();
+            j.waitForCompletion(run);
+
+
+            // Set run back to being incomplete as if persistence failed
+            Field completedField = run.getClass().getDeclaredField("completed");
+            completedField.setAccessible(true);
+            completedField.set(run, false);
+
+            Field resultField = Run.class.getDeclaredField("result");
+            resultField.setAccessible(true);
+            resultField.set(run, null);
+
+            run.save();
+        });
+        story.then( j->{
+            WorkflowJob r = j.jenkins.getItemByFullName(DEFAULT_JOBNAME, WorkflowJob.class);
+            WorkflowRun run = r.getBuildByNumber(build[0]);
+            assertCompletedCleanly(run);
```

Comment:
Maybe add `Assert.assertEquals(Result.SUCCESS, run.getResult());` to show that copying the result from the FlowEndNode works correctly?